### PR TITLE
tests/thread_race: add test for race conditions

### PR DESCRIPTION
**Pull request by ZetaR60**
This adds a test for race conditions in context switching, especially for AVR_CONTEXT_SWAP_TRIGGER on the ATmega.

Tests for problem in #8896

### CODE REVIEW

**Comment by ZetaR60** [comment]
Fixed for #8904

File: tests/thread_race/main.c [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+
+    irq_restore(old_state);
+}
+
+static void _spin(void)
+{
+    /* Volatile so it is not messed with by optimizations */
+    volatile uint8_t i;
+
+    for (i=0; i<255; i++);
```
Comment (kYc0o) [inline tests/thread_race/main.c]: Here and below (for the if-else statements), please follow coding conventions. You may also run uncrustify on this file to correct it automatically.

File: tests/thread_race/main.c [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+}
+
+int main(void)
+{
+    puts("Context swap race condition test application\n");
```
Comment (kYc0o) [inline tests/thread_race/main.c]: Here and below, no need of `\n` in the `puts()` function.

File: tests/thread_race/main.c [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+#include <avr/cpufunc.h>
+#include <avr/interrupt.h>
+#include <stdio.h>
+
+#include "irq.h"
+#include "sched.h"
+#include "thread.h"
+
+/* board.h for AVR_CONTEXT_SWAP_TRIGGER */
+#include "board.h"
```
Comment (kYc0o) [inline tests/thread_race/main.c]: Here and below, I don't like the non-generic nature on this tests, and why AVR should be treated in such special way... I see more convenient to have a generic test which works for all platforms without `#ifdef`s.

If this is due to #8896, please make this PR dependent on #8904, on which `AVR_CONTEXT_SWAP_TRIGGER` doesn't exist anymore.

File: tests/thread_race/main.c [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+    _thread_wake_wo_yield(pid);
+
+#ifdef AVR_CONTEXT_SWAP_TRIGGER
+    AVR_CONTEXT_SWAP_TRIGGER;
+#else
+    thread_yield_higher();
+#endif
+
+    /* Delay so we are not testing for race conditions also */
+    _spin();
```
Comment (kYc0o) [inline tests/thread_race/main.c]: Can you explain a bit better this spin? If I understood correctly, this test is actually used to test race conditions isn't it?

File: tests/thread_race/main.c (hunk 1 above)
Comment (ZetaR60) [reply tests/thread_race/main.c]: Sorry, still new to RIOT conventions. Uncrustified.

File: tests/thread_race/main.c (hunk 2 above)
Comment (ZetaR60) [reply tests/thread_race/main.c]: Removed.

**Comment by ZetaR60** [comment]
Squashed with requested changes incorporated.

**Comment by ZetaR60** [comment]
@kYc0o Squashed. Thanks!

File: tests/thread_race/Makefile [hunk 5]
```diff
@@ -0,0 +1,5 @@
+include ../Makefile.tests_common
+
+DISABLE_MODULE += auto_init
+
```
Comment (kYc0o) [inline tests/thread_race/Makefile]: I forgot this part. I think this test should be run by the CI on hardware. Can you please add:

```makefile
TEST_ON_CI_WHITELIST += all
```

just here?

Comment (ZetaR60) [reply tests/thread_race/Makefile]: Added support for CI hardware testing.

**Comment by kYc0o** [comment]
Removed tests on hardware because unrelated errors block this PR. However the test added by this PR passed.

**Comment by kYc0o** [comment]
Let's go!

**Pull request merged**